asm_to_pil/src/lib.rs

```diff
@@ -24,18 +24,130 @@ use ast::{
 use number::FieldElement;
 
 pub fn compile<T: FieldElement>(analysis: AnalysisASMFile<T>) -> PILGraph<T> {
-    ASMPILConverter::compile(analysis)
+    Session::default().compile(analysis)
 }
 
-struct ConversionOutput<T> {
+#[derive(Clone)]
+struct RelativeObject<T> {
     /// the pil statements for this machine
     pil: Vec<PilStatement<T>>,
     /// the degree of this machine
     degree: u64,
+    /// the relative links to the submachines (one per instruction)
+    relative_links: Vec<UnresolvedLink>,
+    /// the relative submachines (one per submachine)
+    relative_submachines: Vec<SubmachineDeclaration>,
 }
 
-fn compile_machine<T: FieldElement>(input: Machine<T>) -> ConversionOutput<T> {
-    ASMPILConverter::default().convert_machine(input)
+#[derive(Clone)]
+struct UnresolvedLink {
+    instr: Instr,
+    machine_ty: String,
+    loc: String,
+    function_name: String,
+}
+
+#[derive(Default)]
+/// A compilation session
+struct Session<T> {
+    /// the machine types used in the tree
+    machine_types: BTreeMap<String, Machine<T>>,
+    /// the machine types compiled so far
+    cache: BTreeMap<String, RelativeObject<T>>,
+}
+
+impl<T: FieldElement> Session<T> {
+    fn compile_machine(&mut self, ty: String) -> RelativeObject<T> {
+        (*self.cache.entry(ty.clone()).or_insert_with(|| {
+            ASMPILConverter::default().convert_machine(self.machine_types.get(&ty).unwrap().clone())
+        }))
+        .clone()
+    }
+
+    fn compile(&mut self, mut input: AnalysisASMFile<T>) -> PILGraph<T> {
+        let mut objects = BTreeMap::default();
+
+        let location = Location::default().join("main");
+
+        // we start from the main machine
+        let main = match input.machines.len() {
+            // if there is a single machine, treat it as main
+            1 => input.machines.keys().next().unwrap().clone(),
+            // otherwise, use the machine called "Main"
+            _ => {
+                assert!(input.machines.contains_key("Main"));
+                "Main".into()
+            }
+        };
+
+        self.machine_types = std::mem::take(&mut input.machines);
+
+        let mut queue = vec![(location, main)];
+
+        while let Some((absolute_name, ty)) = queue.pop() {
+            let RelativeObject {
+                pil,
+                degree,
+                relative_links,
+                relative_submachines,
+            } = self.compile_machine(ty);
+
+            queue.extend(relative_submachines.into_iter().map(|def| {
```

Comment:
If this machine was already visited before, its relative submachines are already in the queue, and this is adding them again. It doesn't affect soundness because of `objects.insert` at the end, but it causes this whole block to be computed redundantly.

Reply:
Machines are visited only once